codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -407,35 +416,10 @@ public boolean merge(Http2RemoteFlowController.FlowControlled next) {
             // Given that we're merging data into a frame it doesn't really make sense to accumulate padding.
             padding = Math.max(nextData.padding, padding);
             endOfStream = nextData.endOfStream;
-            final CompositeByteBuf compositeByteBuf;
-            if (data instanceof CompositeByteBuf) {
-                compositeByteBuf = (CompositeByteBuf) data;
-            } else {
-                compositeByteBuf = ctx.alloc().compositeBuffer(Integer.MAX_VALUE);
-                compositeByteBuf.addComponent(data);
-                compositeByteBuf.writerIndex(data.readableBytes());
-                data = compositeByteBuf;
-            }
-            compositeByteBuf.addComponent(nextData.data);
-            compositeByteBuf.writerIndex(compositeByteBuf.writerIndex() + nextData.data.readableBytes());
-            size = data.readableBytes() + padding;
-            if (!nextData.promise.isVoid()) {
-                // Replace current promise if void otherwise chain them.
-                if (promise.isVoid()) {
-                    promise = nextData.promise;
-                } else {
-                    promise.addListener(new ChannelFutureListener() {
-                        @Override
-                        public void operationComplete(ChannelFuture future) throws Exception {
-                            if (future.isSuccess()) {
-                                nextData.promise.trySuccess();
-                            } else {
-                                nextData.promise.tryFailure(future.cause());
-                            }
-                        }
-                    });
-                }
-            }
+            bufsAndPromises.addAll(nextData.bufsAndPromises);
+            // Use the next promise for writeComplete
+            this.promise = nextData.promise;
+            size += nextData.size;
```

Comment:
Boundary condition but should we not merge if `this.size() + nextData.Size` (considering padding) would overflow, or exceed the